ddl/ddl_api.go

```diff
@@ -3395,6 +3398,50 @@ func CheckModifyTypeCompatible(origin *types.FieldType, to *types.FieldType) (al
 			msg := fmt.Sprintf("decimal change from decimal(%d, %d) to decimal(%d, %d)", origin.Flen, origin.Decimal, to.Flen, to.Decimal)
 			return msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 		}
+	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
+		switch origin.Tp {
+		case mysql.TypeDuration:
+			switch to.Tp {
+			case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+			}
+		case mysql.TypeDate:
+			switch to.Tp {
+			case mysql.TypeDatetime, mysql.TypeTimestamp:
+				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+			}
+		case mysql.TypeTimestamp:
+			switch to.Tp {
+			case mysql.TypeDuration, mysql.TypeDatetime:
+				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+			}
+		case mysql.TypeDatetime:
+			switch to.Tp {
+			case mysql.TypeTimestamp:
+				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+			}
+		case mysql.TypeYear:
+			switch to.Tp {
+			case mysql.TypeDuration:
+				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+			case mysql.TypeYear:
+			default:
+				return "", ErrTruncatedWrongValue.GenWithStack("banned conversion that must fail")
```

Comment:
eg: 
Take from the year to date as a example, if the row contains the null value or without value, it seems can alter success. So a better way is not to ban it here but actually let cast function to do the judge on the reorg process.

maybe you can do it in the next pr...

Reply:
Got.